<p align="center">
  <img src="https://raw.githubusercontent.com/lacamposm/Diplomado_Metodos_UCentral/main/data/images/imagen_ucentral.jpg" alt="logo_Ucentral" width="400px" height="300px">
</p>

---
# Importación de información.
---
Los datos vienen comunmente en formatos excel, csv, txt, o quizá esté en alguna url. Aunque estos pueden estar en otro formato (JSON, API'S) nos centraremos en los tres primeros.

In [2]:
!pwd

/media/andres-campos/Archivos_Ubuntu/UCentral_Maestria/2023/Diplomado_Metodos_UCentral


In [1]:
!ls data

 APARTAMENTOS.csv	    images
 Athelete_Country_Map.csv   lc_loans.csv
 Athelete_Country_Map.txt   Medals.csv
 Athelete_Sports_Map.csv    mtcars.csv
 CARS2005.SAV		    NOTASBARRAS.xlsx
 downloaded_medals.csv	    state.csv
 downloaded_medals.json     TABLA_PERSONAS.csv
 downloaded_medals.xls	   'Winter Olympic Medals.xlsx'
 espectro.json


In [3]:
## Librerías requeridas
import numpy    as np
import pandas   as pd

## pd.set_option('display.max_columns', None) # Número de columnas máximo a mostrar. None se reemplaza por el # máximo de columnas.

In [4]:
## Función .head() muestra los 5 primeros elementos
df_csv = pd.read_csv('data/downloaded_medals.csv')
df_csv.head()

Unnamed: 0      City      Discipline       Event Event gender   Medal  NOC  \
0           0  Chamonix  Figure skating  individual            M  Silver  AUT   
1           1  Chamonix  Figure skating  individual            W    Gold  AUT   
2           2  Chamonix  Figure skating       pairs            X    Gold  AUT   
3           3  Chamonix       Bobsleigh    four-man            M  Bronze  BEL   
4           4  Chamonix      Ice Hockey  ice hockey            M    Gold  CAN   

        Sport  Year  
0     Skating  1924  
1     Skating  1924  
2     Skating  1924  
3   Bobsleigh  1924  
4  Ice Hockey  1924

In [4]:
df_csv = df_csv.drop(columns=['Unnamed: 0']) ## Borrar columna
df_csv.head()

City      Discipline       Event Event gender   Medal  NOC       Sport  \
0  Chamonix  Figure skating  individual            M  Silver  AUT     Skating   
1  Chamonix  Figure skating  individual            W    Gold  AUT     Skating   
2  Chamonix  Figure skating       pairs            X    Gold  AUT     Skating   
3  Chamonix       Bobsleigh    four-man            M  Bronze  BEL   Bobsleigh   
4  Chamonix      Ice Hockey  ice hockey            M    Gold  CAN  Ice Hockey   

   Year  
0  1924  
1  1924  
2  1924  
3  1924  
4  1924

In [5]:
df_csv.rename(columns={'City':'Ciudades'})

Ciudades      Discipline            Event Event gender   Medal  NOC  \
0     Chamonix  Figure skating       individual            M  Silver  AUT   
1     Chamonix  Figure skating       individual            W    Gold  AUT   
2     Chamonix  Figure skating            pairs            X    Gold  AUT   
3     Chamonix       Bobsleigh         four-man            M  Bronze  BEL   
4     Chamonix      Ice Hockey       ice hockey            M    Gold  CAN   
...        ...             ...              ...          ...     ...  ...   
2306     Turin       Snowboard        Half-pipe            M  Silver  USA   
2307     Turin       Snowboard        Half-pipe            W    Gold  USA   
2308     Turin       Snowboard        Half-pipe            W  Silver  USA   
2309     Turin       Snowboard  Snowboard Cross            M    Gold  USA   
2310     Turin       Snowboard  Snowboard Cross            W  Silver  USA   

           Sport  Year  
0        Skating  1924  
1        Skating  1924  
2        Skating  1924  
3      Bobsleigh  1924  
4     Ice Hockey  1924  
...          ...   ...  
2306      Skiing  2006  
2307      Skiing  2006  
2308      Skiing  2006  
2309      Skiing  2006  
2310      Skiing  2006  

[2311 rows x 8 columns]

In [6]:
## Función .tail() muestra los 5 últimos elementos
df_csv = pd.read_csv('data/downloaded_medals.csv', index_col=0)
df_csv.tail()

City Discipline            Event Event gender   Medal  NOC   Sport  \
2306  Turin  Snowboard        Half-pipe            M  Silver  USA  Skiing   
2307  Turin  Snowboard        Half-pipe            W    Gold  USA  Skiing   
2308  Turin  Snowboard        Half-pipe            W  Silver  USA  Skiing   
2309  Turin  Snowboard  Snowboard Cross            M    Gold  USA  Skiing   
2310  Turin  Snowboard  Snowboard Cross            W  Silver  USA  Skiing   

      Year  
2306  2006  
2307  2006  
2308  2006  
2309  2006  
2310  2006

In [7]:
## Sample
df_csv.sample(frac=0.01)

City       Discipline                         Event  \
2044          Salt Lake City    Speed skating                         1500m   
1740                  Nagano  Cross Country S  Combined 10km + 15km pursuit   
765                  Sapporo   Figure skating                         pairs   
1643                  Nagano          Curling                       curling   
317                     Oslo    Alpine Skiing                        slalom   
1783                  Nagano         Biathlon                 4x7.5km relay   
1101                Sarajevo   Figure skating                   ice dancing   
489                Innsbruck             Luge                       singles   
353        Cortina d'Ampezzo    Alpine Skiing                      downhill   
165   Garmisch-Partenkirchen  Nordic Combined                    individual   
2197                   Turin   Short Track S.                         1500m   
1104                Sarajevo   Figure skating                         pairs   
1444             Lillehammer    Alpine Skiing                        slalom   
1166                 Calgary             Luge                       doubles   
1347             Albertville             Luge                       doubles   
800                Innsbruck  Cross Country S                           5km   
830                Innsbruck      Ski Jumping          K90 individual (70m)   
348        Cortina d'Ampezzo  Cross Country S                          15km   
740                  Sapporo  Cross Country S                  4x10km relay   
1846          Salt Lake City   Short Track S.                         1500m   
440             Squaw Valley         Biathlon                          20km   
1545             Lillehammer    Alpine Skiing               alpine combined   
834                Innsbruck    Alpine Skiing                        slalom   

     Event gender   Medal  NOC     Sport  Year  
2044            M    Gold  USA   Skating  2002  
1740            M  Bronze  KAZ    Skiing  1998  
765             X  Silver  URS   Skating  1972  
1643            M  Silver  CAN   Curling  1998  
317             W    Gold  USA    Skiing  1952  
1783            M  Bronze  RUS  Biathlon  1998  
1101            X  Silver  URS   Skating  1984  
489             W    Gold  EUA      Luge  1964  
353             M  Silver  SUI    Skiing  1956  
165             M    Gold  NOR    Skiing  1936  
2197            W  Silver  KOR   Skating  2006  
1104            X    Gold  URS   Skating  1984  
1444            W  Silver  AUT    Skiing  1994  
1166            X  Silver  GDR      Luge  1988  
1347            X    Gold  GER      Luge  1992  
800             W    Gold  FIN    Skiing  1976  
830             M    Gold  GDR    Skiing  1976  
348             M    Gold  NOR    Skiing  1956  
740             M  Silver  NOR    Skiing  1972  
1846            W  Bronze  BUL   Skating  2002  
440             M  Bronze  URS  Biathlon  1960  
1545            M  Bronze  NOR    Skiing  1994  
834             M  Silver  ITA    Skiing  1976

---
## Archivos excel
---

In [8]:
pip install xlrd

Note: you may need to restart the kernel to use updated packages.


In [9]:
# Archivos Excel
df_Excel= pd.ExcelFile('data/downloaded_medals.xls')
df_Excel.sheet_names

['Sheet1']

In [10]:
df_Excel = df_Excel.parse('Sheet1')
df_Excel.head()

Unnamed: 0      City      Discipline       Event Event gender   Medal  NOC  \
0           0  Chamonix  Figure skating  individual            M  Silver  AUT   
1           1  Chamonix  Figure skating  individual            W    Gold  AUT   
2           2  Chamonix  Figure skating       pairs            X    Gold  AUT   
3           3  Chamonix       Bobsleigh    four-man            M  Bronze  BEL   
4           4  Chamonix      Ice Hockey  ice hockey            M    Gold  CAN   

        Sport  Year  
0     Skating  1924  
1     Skating  1924  
2     Skating  1924  
3   Bobsleigh  1924  
4  Ice Hockey  1924

---
## Archivos JSON
---

In [11]:
df_json = pd.read_json('data/downloaded_medals.json')
df_json.head()

City      Discipline       Event Event gender   Medal  NOC       Sport  \
0  Chamonix  Figure skating  individual            M  Silver  AUT     Skating   
1  Chamonix  Figure skating  individual            W    Gold  AUT     Skating   
2  Chamonix  Figure skating       pairs            X    Gold  AUT     Skating   
3  Chamonix       Bobsleigh    four-man            M  Bronze  BEL   Bobsleigh   
4  Chamonix      Ice Hockey  ice hockey            M    Gold  CAN  Ice Hockey   

   Year  
0  1924  
1  1924  
2  1924  
3  1924  
4  1924

---
### Ejercicio 1
---
Importar el archivo `'Winter Olympic Medals.xlsx'` y definir en cada caso un dataframe para cada hoja. Además en cada caso, mostrar su cabeza y su cola.

**Respuesta:**

In [12]:
#Su código acá

In [13]:
df_Excel_1 = pd.ExcelFile('data/Winter Olympic Medals.xlsx')
df_Excel_1.sheet_names

['Data', 'Ranks', 'Sources']

In [14]:
df_excel = pd.read_excel('data/Winter Olympic Medals.xlsx', sheet_name=['Data', 'Ranks', 'Sources'])
df_aux = df_excel['Data']
df_aux.head()

Year                 Sport                Event      Country Gender  \
0  1924               Bobsled      Men's Four/Five  Switzerland    Men   
1  1924               Bobsled      Men's Four/Five      Britain    Men   
2  1924               Bobsled      Men's Four/Five      Belgium    Men   
3  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   

   Medal Rank   Medal Name of Athlete or Team  Age of Athlete  
0           1    gold           Switzerland-1             NaN  
1           2  silver               Britain-1             NaN  
2           3  bronze               Belgium-1             NaN  
3           1    gold           Thorleif Haug            29.0  
4           2  silver  Johan Grøttumsbraaten            24.0

---
## Desde una URL
---

In [15]:
## Dependerá de donde y como estén los datos.
df_url = pd.read_excel('https://query.data.world/s/zilmgkqzxjhjrfn2pllc4dhikcvrzt')
df_url.head()

Year                 Sport                Event      Country Gender  \
0  1924               Bobsled      Men's Four/Five  Switzerland    Men   
1  1924               Bobsled      Men's Four/Five      Britain    Men   
2  1924               Bobsled      Men's Four/Five      Belgium    Men   
3  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   

   Medal Rank   Medal Name of Athlete or Team  Age of Athlete  
0           1    gold           Switzerland-1             NaN  
1           2  silver               Britain-1             NaN  
2           3  bronze               Belgium-1             NaN  
3           1    gold           Thorleif Haug            29.0  
4           2  silver  Johan Grøttumsbraaten            24.0

In [16]:
url = 'https://raw.githubusercontent.com/joanby/python-ml-course/master/datasets/athletes/Medals.csv'
df_url_1 = pd.read_csv(url, encoding='latin-1')
df_url_1.head()

Athlete   Age  Year Closing Ceremony Date  Gold Medals  \
0    Michael Phelps  23.0  2008            08/24/2008            8   
1    Michael Phelps  19.0  2004            08/29/2004            6   
2    Michael Phelps  27.0  2012            08/12/2012            4   
3  Natalie Coughlin  25.0  2008            08/24/2008            1   
4     Aleksey Nemov  24.0  2000            10/01/2000            2   

   Silver Medals  Bronze Medals  Total Medals  
0              0              0             8  
1              0              2             8  
2              2              0             6  
3              2              3             6  
4              1              3             6

---
## Información del `dataframe`
---

In [17]:
# Copiamos un dataframe
df = df_url.copy()
df.sample(5)

Year                 Sport                  Event        Country Gender  \
537   1964                  Luge        Women's Singles        Germany  Women   
1974  2002       Nordic Combined           Men's Sprint        Finland    Men   
93    1932               Bobsled             Men's Four  United States    Men   
1543  1994      Freestyle Skiing        Women's Aerials         Norway  Women   
2842  2014  Cross-Country Skiing  Women's 30 Kilometers         Norway  Women   

      Medal Rank   Medal Name of Athlete or Team  Age of Athlete  
537            2  silver            Ilse Geisler            23.0  
1974           1    gold          Samppa Lajunen            22.0  
93             1    gold         United States-1             NaN  
1543           3  bronze      Hilde Synnøve Lid            22.0  
2842           2  silver          Therese Johaug            25.0

In [18]:
## Nombres columnas
df.columns

Index(['Year', 'Sport', 'Event', 'Country', 'Gender', 'Medal Rank', 'Medal',
       'Name of Athlete or Team', 'Age of Athlete'],
      dtype='object')

In [19]:
for column in df.columns:
    print(column, type(column))
    print(df[column].head())

Year <class 'str'>
0    1924
1    1924
2    1924
3    1924
4    1924
Name: Year, dtype: int64
Sport <class 'str'>
0                 Bobsled
1                 Bobsled
2                 Bobsled
3    Cross-Country Skiing
4    Cross-Country Skiing
Name: Sport, dtype: object
Event <class 'str'>
0        Men's Four/Five
1        Men's Four/Five
2        Men's Four/Five
3    Men's 18 Kilometers
4    Men's 18 Kilometers
Name: Event, dtype: object
Country <class 'str'>
0    Switzerland
1        Britain
2        Belgium
3         Norway
4         Norway
Name: Country, dtype: object
Gender <class 'str'>
0    Men
1    Men
2    Men
3    Men
4    Men
Name: Gender, dtype: object
Medal Rank <class 'str'>
0    1
1    2
2    3
3    1
4    2
Name: Medal Rank, dtype: int64
Medal <class 'str'>
0      gold
1    silver
2    bronze
3      gold
4    silver
Name: Medal, dtype: object
Name of Athlete or Team <class 'str'>
0             Switzerland-1
1                 Britain-1
2                 Belgium-1
3      

In [20]:
## Tipos de datos en el dataframe
df.dtypes

Year                         int64
Sport                       object
Event                       object
Country                     object
Gender                      object
Medal Rank                   int64
Medal                       object
Name of Athlete or Team     object
Age of Athlete             float64
dtype: object

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2865 entries, 0 to 2864
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Year                     2865 non-null   int64  
 1   Sport                    2865 non-null   object 
 2   Event                    2865 non-null   object 
 3   Country                  2865 non-null   object 
 4   Gender                   2865 non-null   object 
 5   Medal Rank               2865 non-null   int64  
 6   Medal                    2865 non-null   object 
 7   Name of Athlete or Team  2865 non-null   object 
 8   Age of Athlete           2173 non-null   float64
dtypes: float64(1), int64(2), object(6)
memory usage: 201.6+ KB


In [22]:
## pd.DataFrame con las columnas de tipo 'int64'.
df.select_dtypes(include=['int64'])

Year  Medal Rank
0     1924           1
1     1924           2
2     1924           3
3     1924           1
4     1924           2
...    ...         ...
2860  2014           1
2861  2014           2
2862  2014           3
2863  2014           1
2864  2014           2

[2865 rows x 2 columns]

In [23]:
## pd.DataFrame con las columnas numéricas.
df.select_dtypes(np.number)

Year  Medal Rank  Age of Athlete
0     1924           1             NaN
1     1924           2             NaN
2     1924           3             NaN
3     1924           1            29.0
4     1924           2            24.0
...    ...         ...             ...
2860  2014           1            30.0
2861  2014           2            31.0
2862  2014           3            27.0
2863  2014           1             NaN
2864  2014           2             NaN

[2865 rows x 3 columns]

In [24]:
## pd.DataFrame con las columnas tipo object.
df.select_dtypes('object')

Sport                Event      Country Gender   Medal  \
0                  Bobsled      Men's Four/Five  Switzerland    Men    gold   
1                  Bobsled      Men's Four/Five      Britain    Men  silver   
2                  Bobsled      Men's Four/Five      Belgium    Men  bronze   
3     Cross-Country Skiing  Men's 18 Kilometers       Norway    Men    gold   
4     Cross-Country Skiing  Men's 18 Kilometers       Norway    Men  silver   
...                    ...                  ...          ...    ...     ...   
2860  Cross-Country Skiing  Men's 50 Kilometers       Russia    Men    gold   
2861  Cross-Country Skiing  Men's 50 Kilometers       Russia    Men  silver   
2862  Cross-Country Skiing  Men's 50 Kilometers       Russia    Men  bronze   
2863            Ice Hockey     Men's Ice Hockey       Canada    Men    gold   
2864            Ice Hockey     Men's Ice Hockey       Sweden    Men  silver   

     Name of Athlete or Team  
0              Switzerland-1  
1                  Britain-1  
2                  Belgium-1  
3              Thorleif Haug  
4     Johan Grøttumsbraaten  
...                      ...  
2860        Alexander Legkov  
2861       Maxim Vylegzhanin  
2862         Ilia Chernousov  
2863                  Canada  
2864                  Sweden  

[2865 rows x 6 columns]

In [25]:
df.describe().T # df.describe.transpose()

count         mean        std     min     25%     50%  \
Year            2865.0  1986.488656  23.489775  1924.0  1972.0  1992.0   
Medal Rank      2865.0     1.995812   0.815773     1.0     1.0     2.0   
Age of Athlete  2173.0    25.152324   4.271764    14.0    22.0    25.0   

                   75%     max  
Year            2006.0  2014.0  
Medal Rank         3.0     3.0  
Age of Athlete    28.0    42.0

In [26]:
df.describe(include=['object']).T # df.describe.transpose()

count unique            top  freq
Sport                    2865     15   Speedskating   527
Event                    2865    127  Men's Singles   108
Country                  2865     44         Norway   329
Gender                   2865      3            Men  1711
Medal                    2865      3           gold   959
Name of Athlete or Team  2865   1430         Canada    49

In [27]:
df

Year                 Sport                Event      Country Gender  \
0     1924               Bobsled      Men's Four/Five  Switzerland    Men   
1     1924               Bobsled      Men's Four/Five      Britain    Men   
2     1924               Bobsled      Men's Four/Five      Belgium    Men   
3     1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4     1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
...    ...                   ...                  ...          ...    ...   
2860  2014  Cross-Country Skiing  Men's 50 Kilometers       Russia    Men   
2861  2014  Cross-Country Skiing  Men's 50 Kilometers       Russia    Men   
2862  2014  Cross-Country Skiing  Men's 50 Kilometers       Russia    Men   
2863  2014            Ice Hockey     Men's Ice Hockey       Canada    Men   
2864  2014            Ice Hockey     Men's Ice Hockey       Sweden    Men   

      Medal Rank   Medal Name of Athlete or Team  Age of Athlete  
0              1    gold           Switzerland-1             NaN  
1              2  silver               Britain-1             NaN  
2              3  bronze               Belgium-1             NaN  
3              1    gold           Thorleif Haug            29.0  
4              2  silver  Johan Grøttumsbraaten            24.0  
...          ...     ...                     ...             ...  
2860           1    gold        Alexander Legkov            30.0  
2861           2  silver       Maxim Vylegzhanin            31.0  
2862           3  bronze         Ilia Chernousov            27.0  
2863           1    gold                  Canada             NaN  
2864           2  silver                  Sweden             NaN  

[2865 rows x 9 columns]

In [28]:
df = df.rename(columns={'Event':'Evento', 'Gender':'Género', 'Medal':'Medalla',
                        'Age of Athlete':'Edad_Atleta',
                        'Name of Athlete or Team':'Nombre_atleta_o_Equipo'
                        })
df.head()

Year                 Sport               Evento      Country Género  \
0  1924               Bobsled      Men's Four/Five  Switzerland    Men   
1  1924               Bobsled      Men's Four/Five      Britain    Men   
2  1924               Bobsled      Men's Four/Five      Belgium    Men   
3  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   

   Medal Rank Medalla  Nombre_atleta_o_Equipo  Edad_Atleta  
0           1    gold           Switzerland-1          NaN  
1           2  silver               Britain-1          NaN  
2           3  bronze               Belgium-1          NaN  
3           1    gold           Thorleif Haug         29.0  
4           2  silver  Johan Grøttumsbraaten         24.0

---
## Sub dataframes
---

In [29]:
## Serie de pandas
display(df['Edad_Atleta'])
type(df['Edad_Atleta'])

0        NaN
1        NaN
2        NaN
3       29.0
4       24.0
        ... 
2860    30.0
2861    31.0
2862    27.0
2863     NaN
2864     NaN
Name: Edad_Atleta, Length: 2865, dtype: float64

pandas.core.series.Series

In [30]:
display(df['Evento'])
### Otra forma de obtener la columna (pd.Series)
df.Evento

0           Men's Four/Five
1           Men's Four/Five
2           Men's Four/Five
3       Men's 18 Kilometers
4       Men's 18 Kilometers
               ...         
2860    Men's 50 Kilometers
2861    Men's 50 Kilometers
2862    Men's 50 Kilometers
2863       Men's Ice Hockey
2864       Men's Ice Hockey
Name: Evento, Length: 2865, dtype: object

0           Men's Four/Five
1           Men's Four/Five
2           Men's Four/Five
3       Men's 18 Kilometers
4       Men's 18 Kilometers
               ...         
2860    Men's 50 Kilometers
2861    Men's 50 Kilometers
2862    Men's 50 Kilometers
2863       Men's Ice Hockey
2864       Men's Ice Hockey
Name: Evento, Length: 2865, dtype: object

In [31]:
df['Género'].unique()

array(['Men', 'Mixed', 'Women'], dtype=object)

In [32]:
subset = df[['Year', 'Sport', 'Evento']]
subset

Year                 Sport               Evento
0     1924               Bobsled      Men's Four/Five
1     1924               Bobsled      Men's Four/Five
2     1924               Bobsled      Men's Four/Five
3     1924  Cross-Country Skiing  Men's 18 Kilometers
4     1924  Cross-Country Skiing  Men's 18 Kilometers
...    ...                   ...                  ...
2860  2014  Cross-Country Skiing  Men's 50 Kilometers
2861  2014  Cross-Country Skiing  Men's 50 Kilometers
2862  2014  Cross-Country Skiing  Men's 50 Kilometers
2863  2014            Ice Hockey     Men's Ice Hockey
2864  2014            Ice Hockey     Men's Ice Hockey

[2865 rows x 3 columns]

In [33]:
df['Sport'].value_counts()

Speedskating                527
Cross-Country Skiing        471
Alpine Skiing               428
Figure Skating              235
Biathlon                    225
Short-Track Speedskating    144
Ski Jumping                 135
Bobsled                     133
Luge                        129
Nordic Combined             102
Freestyle Skiing            102
Snowboarding                 90
Ice Hockey                   81
Curling                      33
Skeleton                     30
Name: Sport, dtype: int64

In [34]:
df['Sport'].value_counts(normalize=True, dropna=False)

Speedskating                0.183944
Cross-Country Skiing        0.164398
Alpine Skiing               0.149389
Figure Skating              0.082024
Biathlon                    0.078534
Short-Track Speedskating    0.050262
Ski Jumping                 0.047120
Bobsled                     0.046422
Luge                        0.045026
Nordic Combined             0.035602
Freestyle Skiing            0.035602
Snowboarding                0.031414
Ice Hockey                  0.028272
Curling                     0.011518
Skeleton                    0.010471
Name: Sport, dtype: float64

In [35]:
df['Género'].value_counts(normalize=True, dropna=False)*100

Men      59.720768
Women    36.474695
Mixed     3.804538
Name: Género, dtype: float64

---
## `.loc` y `.iloc`
---

Si deseamos seleccionar un subdataframe, con registros y columnas particulares podemos hacerlos con estas 2 funciones. `.loc` es la localización en texto y `.iloc()` será vía indexación numérica.

In [36]:
df.head()

Year                 Sport               Evento      Country Género  \
0  1924               Bobsled      Men's Four/Five  Switzerland    Men   
1  1924               Bobsled      Men's Four/Five      Britain    Men   
2  1924               Bobsled      Men's Four/Five      Belgium    Men   
3  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   

   Medal Rank Medalla  Nombre_atleta_o_Equipo  Edad_Atleta  
0           1    gold           Switzerland-1          NaN  
1           2  silver               Britain-1          NaN  
2           3  bronze               Belgium-1          NaN  
3           1    gold           Thorleif Haug         29.0  
4           2  silver  Johan Grøttumsbraaten         24.0

In [37]:
df['Sport'][4]

'Cross-Country Skiing'

In [38]:
df.loc[4] ## Registro 4 de la tabla.

Year                                        1924
Sport                       Cross-Country Skiing
Evento                       Men's 18 Kilometers
Country                                   Norway
Género                                       Men
Medal Rank                                     2
Medalla                                   silver
Nombre_atleta_o_Equipo    Johan Grøttumsbraaten
Edad_Atleta                                 24.0
Name: 4, dtype: object

In [39]:
type(df.loc[4])

pandas.core.series.Series

In [40]:
df.loc[50:53] ## 8 Primeros registros del dataframe

Year                 Sport               Evento        Country Género  \
50  1928               Bobsled      Men's Four/Five  United States    Men   
51  1928               Bobsled      Men's Four/Five        Germany    Men   
52  1928  Cross-Country Skiing  Men's 18 Kilometers         Norway    Men   
53  1928  Cross-Country Skiing  Men's 18 Kilometers         Norway    Men   

    Medal Rank Medalla  Nombre_atleta_o_Equipo  Edad_Atleta  
50           2  silver         United States-1          NaN  
51           3  bronze               Germany-2          NaN  
52           1    gold  Johan Grøttumsbraaten         28.0  
53           2  silver               Ole Hegge         29.0

In [41]:
df.iloc[10:20] # Registros desde el 11 y hasta el 20

Year           Sport           Evento        Country Género  Medal Rank  \
10  1924         Curling    Men's Curling         Sweden    Men           2   
11  1924         Curling    Men's Curling         France    Men           3   
12  1924  Figure Skating    Men's Singles         Sweden    Men           1   
13  1924  Figure Skating    Men's Singles        Austria    Men           2   
14  1924  Figure Skating    Men's Singles    Switzerland    Men           3   
15  1924  Figure Skating      Mixed Pairs        Austria  Mixed           1   
16  1924  Figure Skating      Mixed Pairs        Finland  Mixed           2   
17  1924  Figure Skating      Mixed Pairs         France  Mixed           3   
18  1924  Figure Skating  Women's Singles        Austria  Women           1   
19  1924  Figure Skating  Women's Singles  United States  Women           2   

   Medalla Nombre_atleta_o_Equipo  Edad_Atleta  
10  silver                 Sweden          NaN  
11  bronze                 France          NaN  
12    gold      Gillis Grafström         30.0  
13  silver           Willy Böckl         30.0  
14  bronze       Georges Gautschi         19.0  
15    gold                Austria          NaN  
16  silver                Finland          NaN  
17  bronze               France-1          NaN  
18    gold     Herma Planck-Szabo         21.0  
19  silver       Beatrix Loughran         23.0

In [42]:
len(df['Year'])

2865

In [43]:
pares = []
for i in range(len(df['Year'])):
    if i%2==0:
        pares.append(i)

pares

[0,
 2,
 4,
 6,
 8,
 10,
 12,
 14,
 16,
 18,
 20,
 22,
 24,
 26,
 28,
 30,
 32,
 34,
 36,
 38,
 40,
 42,
 44,
 46,
 48,
 50,
 52,
 54,
 56,
 58,
 60,
 62,
 64,
 66,
 68,
 70,
 72,
 74,
 76,
 78,
 80,
 82,
 84,
 86,
 88,
 90,
 92,
 94,
 96,
 98,
 100,
 102,
 104,
 106,
 108,
 110,
 112,
 114,
 116,
 118,
 120,
 122,
 124,
 126,
 128,
 130,
 132,
 134,
 136,
 138,
 140,
 142,
 144,
 146,
 148,
 150,
 152,
 154,
 156,
 158,
 160,
 162,
 164,
 166,
 168,
 170,
 172,
 174,
 176,
 178,
 180,
 182,
 184,
 186,
 188,
 190,
 192,
 194,
 196,
 198,
 200,
 202,
 204,
 206,
 208,
 210,
 212,
 214,
 216,
 218,
 220,
 222,
 224,
 226,
 228,
 230,
 232,
 234,
 236,
 238,
 240,
 242,
 244,
 246,
 248,
 250,
 252,
 254,
 256,
 258,
 260,
 262,
 264,
 266,
 268,
 270,
 272,
 274,
 276,
 278,
 280,
 282,
 284,
 286,
 288,
 290,
 292,
 294,
 296,
 298,
 300,
 302,
 304,
 306,
 308,
 310,
 312,
 314,
 316,
 318,
 320,
 322,
 324,
 326,
 328,
 330,
 332,
 334,
 336,
 338,
 340,
 342,
 344,
 346,
 348,
 350,

In [44]:
pares = [i for i in range(len(df['Year'])) if i%2==0]
df.loc[pares]

Year                 Sport               Evento      Country Género  \
0     1924               Bobsled      Men's Four/Five  Switzerland    Men   
2     1924               Bobsled      Men's Four/Five      Belgium    Men   
4     1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
6     1924  Cross-Country Skiing  Men's 50 Kilometers       Norway    Men   
8     1924  Cross-Country Skiing  Men's 50 Kilometers       Norway    Men   
...    ...                   ...                  ...          ...    ...   
2856  2014          Speedskating   Men's Team Pursuit       Poland    Men   
2858  2014               Bobsled           Men's Four       Latvia    Men   
2860  2014  Cross-Country Skiing  Men's 50 Kilometers       Russia    Men   
2862  2014  Cross-Country Skiing  Men's 50 Kilometers       Russia    Men   
2864  2014            Ice Hockey     Men's Ice Hockey       Sweden    Men   

      Medal Rank Medalla  Nombre_atleta_o_Equipo  Edad_Atleta  
0              1    gold           Switzerland-1          NaN  
2              3  bronze               Belgium-1          NaN  
4              2  silver  Johan Grøttumsbraaten         24.0  
6              1    gold           Thorleif Haug         29.0  
8              3  bronze  Johan Grøttumsbraaten         24.0  
...          ...     ...                     ...          ...  
2856           3  bronze                  Poland          NaN  
2858           2  silver                  Latvia          NaN  
2860           1    gold        Alexander Legkov         30.0  
2862           3  bronze         Ilia Chernousov         27.0  
2864           2  silver                  Sweden          NaN  

[1433 rows x 9 columns]

In [45]:
impares = [i for i in range(len(df['Year'])) if i%2==1]
df.iloc[impares]

Year                 Sport               Evento        Country Género  \
1     1924               Bobsled      Men's Four/Five        Britain    Men   
3     1924  Cross-Country Skiing  Men's 18 Kilometers         Norway    Men   
5     1924  Cross-Country Skiing  Men's 18 Kilometers        Finland    Men   
7     1924  Cross-Country Skiing  Men's 50 Kilometers         Norway    Men   
9     1924               Curling        Men's Curling        Britain    Men   
...    ...                   ...                  ...            ...    ...   
2855  2014          Speedskating   Men's Team Pursuit    South Korea    Men   
2857  2014               Bobsled           Men's Four         Russia    Men   
2859  2014               Bobsled           Men's Four  United States    Men   
2861  2014  Cross-Country Skiing  Men's 50 Kilometers         Russia    Men   
2863  2014            Ice Hockey     Men's Ice Hockey         Canada    Men   

      Medal Rank Medalla Nombre_atleta_o_Equipo  Edad_Atleta  
1              2  silver              Britain-1          NaN  
3              1    gold          Thorleif Haug         29.0  
5              3  bronze            Tapani Niku         28.0  
7              2  silver     Thoralf Strømstad         27.0  
9              1    gold                Britain          NaN  
...          ...     ...                    ...          ...  
2855           2  silver            South Korea          NaN  
2857           1    gold                 Russia          NaN  
2859           3  bronze          United States          NaN  
2861           2  silver      Maxim Vylegzhanin         31.0  
2863           1    gold                 Canada          NaN  

[1432 rows x 9 columns]

In [46]:
df[['Sport','Evento']]

Sport               Evento
0                  Bobsled      Men's Four/Five
1                  Bobsled      Men's Four/Five
2                  Bobsled      Men's Four/Five
3     Cross-Country Skiing  Men's 18 Kilometers
4     Cross-Country Skiing  Men's 18 Kilometers
...                    ...                  ...
2860  Cross-Country Skiing  Men's 50 Kilometers
2861  Cross-Country Skiing  Men's 50 Kilometers
2862  Cross-Country Skiing  Men's 50 Kilometers
2863            Ice Hockey     Men's Ice Hockey
2864            Ice Hockey     Men's Ice Hockey

[2865 rows x 2 columns]

In [47]:
df.loc[:,['Sport','Evento']]

Sport               Evento
0                  Bobsled      Men's Four/Five
1                  Bobsled      Men's Four/Five
2                  Bobsled      Men's Four/Five
3     Cross-Country Skiing  Men's 18 Kilometers
4     Cross-Country Skiing  Men's 18 Kilometers
...                    ...                  ...
2860  Cross-Country Skiing  Men's 50 Kilometers
2861  Cross-Country Skiing  Men's 50 Kilometers
2862  Cross-Country Skiing  Men's 50 Kilometers
2863            Ice Hockey     Men's Ice Hockey
2864            Ice Hockey     Men's Ice Hockey

[2865 rows x 2 columns]

In [48]:
for i,j in enumerate(df.columns):
  print((i,j))

(0, 'Year')
(1, 'Sport')
(2, 'Evento')
(3, 'Country')
(4, 'Género')
(5, 'Medal Rank')
(6, 'Medalla')
(7, 'Nombre_atleta_o_Equipo')
(8, 'Edad_Atleta')


In [49]:
df.iloc[pares, [0,1,7]]

Year                 Sport  Nombre_atleta_o_Equipo
0     1924               Bobsled           Switzerland-1
2     1924               Bobsled               Belgium-1
4     1924  Cross-Country Skiing  Johan Grøttumsbraaten
6     1924  Cross-Country Skiing           Thorleif Haug
8     1924  Cross-Country Skiing  Johan Grøttumsbraaten
...    ...                   ...                     ...
2856  2014          Speedskating                  Poland
2858  2014               Bobsled                  Latvia
2860  2014  Cross-Country Skiing        Alexander Legkov
2862  2014  Cross-Country Skiing         Ilia Chernousov
2864  2014            Ice Hockey                  Sweden

[1433 rows x 3 columns]

---
## Máscaras
---
Supongamos que deseamos obtener los regístros que cumplen cierta condición. Por ejemplo, todos aquellos donde el género sea mujer, las personas menores de 30 años, aquellos que ganaron plata, o quizá mujeres que ganaron plata y son menores de 30 años.

Para hacer esto necesitamos una serie de `pandas` de tipo booleano, que determina que registros cumplen la condición deseada.

In [50]:
df.head()

Year                 Sport               Evento      Country Género  \
0  1924               Bobsled      Men's Four/Five  Switzerland    Men   
1  1924               Bobsled      Men's Four/Five      Britain    Men   
2  1924               Bobsled      Men's Four/Five      Belgium    Men   
3  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   

   Medal Rank Medalla  Nombre_atleta_o_Equipo  Edad_Atleta  
0           1    gold           Switzerland-1          NaN  
1           2  silver               Britain-1          NaN  
2           3  bronze               Belgium-1          NaN  
3           1    gold           Thorleif Haug         29.0  
4           2  silver  Johan Grøttumsbraaten         24.0

In [51]:
mask = (df['Year'] > 1950)
mask

0       False
1       False
2       False
3       False
4       False
        ...  
2860     True
2861     True
2862     True
2863     True
2864     True
Name: Year, Length: 2865, dtype: bool

In [52]:
## Tenemos todos los registros de quienes ganaron medalla después del Año 1948
df[mask]

Year                 Sport               Evento  Country Género  \
251   1952         Alpine Skiing       Men's Downhill    Italy    Men   
252   1952         Alpine Skiing       Men's Downhill  Austria    Men   
253   1952         Alpine Skiing       Men's Downhill  Austria    Men   
254   1952         Alpine Skiing   Men's Giant Slalom   Norway    Men   
255   1952         Alpine Skiing   Men's Giant Slalom  Austria    Men   
...    ...                   ...                  ...      ...    ...   
2860  2014  Cross-Country Skiing  Men's 50 Kilometers   Russia    Men   
2861  2014  Cross-Country Skiing  Men's 50 Kilometers   Russia    Men   
2862  2014  Cross-Country Skiing  Men's 50 Kilometers   Russia    Men   
2863  2014            Ice Hockey     Men's Ice Hockey   Canada    Men   
2864  2014            Ice Hockey     Men's Ice Hockey   Sweden    Men   

      Medal Rank Medalla Nombre_atleta_o_Equipo  Edad_Atleta  
251            1    gold             Zeno Colò         31.0  
252            2  silver       Othmar Schneider         23.0  
253            3  bronze       Christian Pravda         24.0  
254            1    gold          Stein Eriksen         24.0  
255            2  silver       Christian Pravda         24.0  
...          ...     ...                    ...          ...  
2860           1    gold       Alexander Legkov         30.0  
2861           2  silver      Maxim Vylegzhanin         31.0  
2862           3  bronze        Ilia Chernousov         27.0  
2863           1    gold                 Canada          NaN  
2864           2  silver                 Sweden          NaN  

[2614 rows x 9 columns]

In [53]:
mask_1 = (df['Edad_Atleta'] <= 20)

In [54]:
df_menores_de_20 = df[mask_1]
display(df_menores_de_20.describe().T)
df_menores_de_20.head()

count         mean        std     min     25%     50%     75%  \
Year         299.0  1983.471572  22.087115  1924.0  1968.0  1984.0  2002.0   
Medal Rank   299.0     2.023411   0.812726     1.0     1.0     2.0     3.0   
Edad_Atleta  299.0    18.856187   1.332123    14.0    18.0    19.0    20.0   

                max  
Year         2014.0  
Medal Rank      3.0  
Edad_Atleta    20.0

Year           Sport           Evento        Country Género  Medal Rank  \
14  1924  Figure Skating    Men's Singles    Switzerland    Men           3   
60  1928  Figure Skating    Men's Singles        Belgium    Men           3   
64  1928  Figure Skating  Women's Singles         Norway  Women           1   
65  1928  Figure Skating  Women's Singles        Austria  Women           2   
74  1928        Skeleton   Men's Skeleton  United States    Men           2   

   Medalla Nombre_atleta_o_Equipo  Edad_Atleta  
14  bronze       Georges Gautschi         19.0  
60  bronze   Robert Van Zeebroeck         18.0  
64    gold            Sonja Henie         15.0  
65  silver          Fritzi Burger         17.0  
74  silver            Jack Heaton         19.0

In [55]:
df_menores_de_20.reset_index(drop=True, inplace=True)
df_menores_de_20

Year                     Sport                Evento        Country  \
0    1924            Figure Skating         Men's Singles    Switzerland   
1    1928            Figure Skating         Men's Singles        Belgium   
2    1928            Figure Skating       Women's Singles         Norway   
3    1928            Figure Skating       Women's Singles        Austria   
4    1928                  Skeleton        Men's Skeleton  United States   
..    ...                       ...                   ...            ...   
294  2014             Alpine Skiing        Women's Slalom  United States   
295  2014  Short-Track Speedskating  Women's 1,000 Meters          China   
296  2014  Short-Track Speedskating  Women's 1,000 Meters    South Korea   
297  2014  Short-Track Speedskating      Men's 500 Meters          China   
298  2014             Alpine Skiing          Men's Slalom         Norway   

    Género  Medal Rank Medalla Nombre_atleta_o_Equipo  Edad_Atleta  
0      Men           3  bronze       Georges Gautschi         19.0  
1      Men           3  bronze   Robert Van Zeebroeck         18.0  
2    Women           1    gold            Sonja Henie         15.0  
3    Women           2  silver          Fritzi Burger         17.0  
4      Men           2  silver            Jack Heaton         19.0  
..     ...         ...     ...                    ...          ...  
294  Women           1    gold       Mikaela Shiffrin         18.0  
295  Women           2  silver              Kexin Fan         20.0  
296  Women           3  bronze           Suk Hee Shim         17.0  
297    Men           2  silver              Dajing Wu         19.0  
298    Men           3  bronze   Henrik Kristoffersen         19.0  

[299 rows x 9 columns]

In [56]:
## Podemos juntar ambas condiciones
df[mask & mask_1]#.describe().T

Year                     Sport                Evento        Country  \
262   1952             Alpine Skiing      Women's Downhill          Italy   
263   1952             Alpine Skiing  Women's Giant Slalom  United States   
266   1952             Alpine Skiing        Women's Slalom  United States   
289   1952            Figure Skating         Men's Singles  United States   
294   1952            Figure Skating       Women's Singles  United States   
...    ...                       ...                   ...            ...   
2814  2014             Alpine Skiing        Women's Slalom  United States   
2827  2014  Short-Track Speedskating  Women's 1,000 Meters          China   
2828  2014  Short-Track Speedskating  Women's 1,000 Meters    South Korea   
2830  2014  Short-Track Speedskating      Men's 500 Meters          China   
2837  2014             Alpine Skiing          Men's Slalom         Norway   

     Género  Medal Rank Medalla   Nombre_atleta_o_Equipo  Edad_Atleta  
262   Women           3  bronze  Giuliana Chenal-Minuzzo         20.0  
263   Women           1    gold     Andrea Mead-Lawrence         19.0  
266   Women           1    gold     Andrea Mead-Lawrence         19.0  
289     Men           3  bronze               Jim Grogan         20.0  
294   Women           2  silver          Tenley Albright         16.0  
...     ...         ...     ...                      ...          ...  
2814  Women           1    gold         Mikaela Shiffrin         18.0  
2827  Women           2  silver                Kexin Fan         20.0  
2828  Women           3  bronze             Suk Hee Shim         17.0  
2830    Men           2  silver                Dajing Wu         19.0  
2837    Men           3  bronze     Henrik Kristoffersen         19.0  

[276 rows x 9 columns]

In [57]:
df['Medalla'].value_counts()

gold      959
silver    958
bronze    948
Name: Medalla, dtype: int64

In [58]:
## informació para un segmento particular.
df[mask & mask_1]['Medalla'].value_counts()

bronze    94
silver    93
gold      89
Name: Medalla, dtype: int64

In [59]:
mask_2 = df['Country'] == 'China'
df[mask_1 & mask_1 & mask_2].reset_index(drop=True)['Género'].value_counts(normalize=True)*100

Women    75.0
Men      25.0
Name: Género, dtype: float64

---
## NaN
---
Suele ocurrir que en nuestro `dataframe` se tengan valores perdidos. Estos deben ser tratados por métodos de imputación que mas adelante discutiremos. Por ahora veremos como eliminarlos o modificarlos.

In [60]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2865 entries, 0 to 2864
Data columns (total 9 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Year                    2865 non-null   int64  
 1   Sport                   2865 non-null   object 
 2   Evento                  2865 non-null   object 
 3   Country                 2865 non-null   object 
 4   Género                  2865 non-null   object 
 5   Medal Rank              2865 non-null   int64  
 6   Medalla                 2865 non-null   object 
 7   Nombre_atleta_o_Equipo  2865 non-null   object 
 8   Edad_Atleta             2173 non-null   float64
dtypes: float64(1), int64(2), object(6)
memory usage: 201.6+ KB


In [61]:
df_drop = df.dropna()
df_drop.head()

Year                 Sport               Evento  Country Género  \
3  1924  Cross-Country Skiing  Men's 18 Kilometers   Norway    Men   
4  1924  Cross-Country Skiing  Men's 18 Kilometers   Norway    Men   
5  1924  Cross-Country Skiing  Men's 18 Kilometers  Finland    Men   
6  1924  Cross-Country Skiing  Men's 50 Kilometers   Norway    Men   
7  1924  Cross-Country Skiing  Men's 50 Kilometers   Norway    Men   

   Medal Rank Medalla  Nombre_atleta_o_Equipo  Edad_Atleta  
3           1    gold           Thorleif Haug         29.0  
4           2  silver  Johan Grøttumsbraaten         24.0  
5           3  bronze             Tapani Niku         28.0  
6           1    gold           Thorleif Haug         29.0  
7           2  silver      Thoralf Strømstad         27.0

In [62]:
df['Edad_Atleta'].isna()

0        True
1        True
2        True
3       False
4       False
        ...  
2860    False
2861    False
2862    False
2863     True
2864     True
Name: Edad_Atleta, Length: 2865, dtype: bool

In [63]:
df[df['Edad_Atleta'].isna()]

Year       Sport            Evento        Country Género  Medal Rank  \
0     1924     Bobsled   Men's Four/Five    Switzerland    Men           1   
1     1924     Bobsled   Men's Four/Five        Britain    Men           2   
2     1924     Bobsled   Men's Four/Five        Belgium    Men           3   
9     1924     Curling     Men's Curling        Britain    Men           1   
10    1924     Curling     Men's Curling         Sweden    Men           2   
...    ...         ...               ...            ...    ...         ...   
2857  2014     Bobsled        Men's Four         Russia    Men           1   
2858  2014     Bobsled        Men's Four         Latvia    Men           2   
2859  2014     Bobsled        Men's Four  United States    Men           3   
2863  2014  Ice Hockey  Men's Ice Hockey         Canada    Men           1   
2864  2014  Ice Hockey  Men's Ice Hockey         Sweden    Men           2   

     Medalla Nombre_atleta_o_Equipo  Edad_Atleta  
0       gold          Switzerland-1          NaN  
1     silver              Britain-1          NaN  
2     bronze              Belgium-1          NaN  
9       gold                Britain          NaN  
10    silver                 Sweden          NaN  
...      ...                    ...          ...  
2857    gold                 Russia          NaN  
2858  silver                 Latvia          NaN  
2859  bronze          United States          NaN  
2863    gold                 Canada          NaN  
2864  silver                 Sweden          NaN  

[692 rows x 9 columns]

In [64]:
## Podemos obtener la negación de la máscara
df[~(df['Edad_Atleta'].isna())]

Year                 Sport                 Evento   Country Género  \
3     1924  Cross-Country Skiing    Men's 18 Kilometers    Norway    Men   
4     1924  Cross-Country Skiing    Men's 18 Kilometers    Norway    Men   
5     1924  Cross-Country Skiing    Men's 18 Kilometers   Finland    Men   
6     1924  Cross-Country Skiing    Men's 50 Kilometers    Norway    Men   
7     1924  Cross-Country Skiing    Men's 50 Kilometers    Norway    Men   
...    ...                   ...                    ...       ...    ...   
2849  2014          Snowboarding  Men's Parallel Slalom  Slovenia    Men   
2850  2014          Snowboarding  Men's Parallel Slalom   Austria    Men   
2860  2014  Cross-Country Skiing    Men's 50 Kilometers    Russia    Men   
2861  2014  Cross-Country Skiing    Men's 50 Kilometers    Russia    Men   
2862  2014  Cross-Country Skiing    Men's 50 Kilometers    Russia    Men   

      Medal Rank Medalla  Nombre_atleta_o_Equipo  Edad_Atleta  
3              1    gold           Thorleif Haug         29.0  
4              2  silver  Johan Grøttumsbraaten         24.0  
5              3  bronze             Tapani Niku         28.0  
6              1    gold           Thorleif Haug         29.0  
7              2  silver      Thoralf Strømstad         27.0  
...          ...     ...                     ...          ...  
2849           2  silver               Zan Kosir         29.0  
2850           3  bronze           Benjamin Karl         28.0  
2860           1    gold        Alexander Legkov         30.0  
2861           2  silver       Maxim Vylegzhanin         31.0  
2862           3  bronze         Ilia Chernousov         27.0  

[2173 rows x 9 columns]

In [65]:
## Todo regístro None o NaN es reemplazado por 0
df.fillna(0)

Year                 Sport               Evento      Country Género  \
0     1924               Bobsled      Men's Four/Five  Switzerland    Men   
1     1924               Bobsled      Men's Four/Five      Britain    Men   
2     1924               Bobsled      Men's Four/Five      Belgium    Men   
3     1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4     1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
...    ...                   ...                  ...          ...    ...   
2860  2014  Cross-Country Skiing  Men's 50 Kilometers       Russia    Men   
2861  2014  Cross-Country Skiing  Men's 50 Kilometers       Russia    Men   
2862  2014  Cross-Country Skiing  Men's 50 Kilometers       Russia    Men   
2863  2014            Ice Hockey     Men's Ice Hockey       Canada    Men   
2864  2014            Ice Hockey     Men's Ice Hockey       Sweden    Men   

      Medal Rank Medalla  Nombre_atleta_o_Equipo  Edad_Atleta  
0              1    gold           Switzerland-1          0.0  
1              2  silver               Britain-1          0.0  
2              3  bronze               Belgium-1          0.0  
3              1    gold           Thorleif Haug         29.0  
4              2  silver  Johan Grøttumsbraaten         24.0  
...          ...     ...                     ...          ...  
2860           1    gold        Alexander Legkov         30.0  
2861           2  silver       Maxim Vylegzhanin         31.0  
2862           3  bronze         Ilia Chernousov         27.0  
2863           1    gold                  Canada          0.0  
2864           2  silver                  Sweden          0.0  

[2865 rows x 9 columns]

In [66]:
df['Edad_Atleta'].fillna(0)

0        0.0
1        0.0
2        0.0
3       29.0
4       24.0
        ... 
2860    30.0
2861    31.0
2862    27.0
2863     0.0
2864     0.0
Name: Edad_Atleta, Length: 2865, dtype: float64

In [67]:
df['Edad_Atleta_imputado'] = df['Edad_Atleta'].fillna(df['Edad_Atleta'].mean())
df.head(10)

Year                 Sport               Evento      Country Género  \
0  1924               Bobsled      Men's Four/Five  Switzerland    Men   
1  1924               Bobsled      Men's Four/Five      Britain    Men   
2  1924               Bobsled      Men's Four/Five      Belgium    Men   
3  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
4  1924  Cross-Country Skiing  Men's 18 Kilometers       Norway    Men   
5  1924  Cross-Country Skiing  Men's 18 Kilometers      Finland    Men   
6  1924  Cross-Country Skiing  Men's 50 Kilometers       Norway    Men   
7  1924  Cross-Country Skiing  Men's 50 Kilometers       Norway    Men   
8  1924  Cross-Country Skiing  Men's 50 Kilometers       Norway    Men   
9  1924               Curling        Men's Curling      Britain    Men   

   Medal Rank Medalla  Nombre_atleta_o_Equipo  Edad_Atleta  \
0           1    gold           Switzerland-1          NaN   
1           2  silver               Britain-1          NaN   
2           3  bronze               Belgium-1          NaN   
3           1    gold           Thorleif Haug         29.0   
4           2  silver  Johan Grøttumsbraaten         24.0   
5           3  bronze             Tapani Niku         28.0   
6           1    gold           Thorleif Haug         29.0   
7           2  silver      Thoralf Strømstad         27.0   
8           3  bronze  Johan Grøttumsbraaten         24.0   
9           1    gold                 Britain          NaN   

   Edad_Atleta_imputado  
0             25.152324  
1             25.152324  
2             25.152324  
3             29.000000  
4             24.000000  
5             28.000000  
6             29.000000  
7             27.000000  
8             24.000000  
9             25.152324

In [68]:
df_drop.shape

(2173, 9)

---
## `pandas.crosstab`
---
La función `pandas.crosstab` devuelve la tabla de contingencia resultante de cruzar dos o más columnas de un dataframe. Por defecto ell resultado evalúa las frecuencias (absolutas o relativas) de cada combinación de valores, es posible especificar una función de agregación.

In [69]:
pd.crosstab(df['Género'],df['Medalla'])

Medalla  bronze  gold  silver
Género                       
Men         568   573     570
Mixed        36    37      36
Women       344   349     352

In [70]:
## Podemos obtener la suma por filas y columnas (subtotales) y renombrar esta
pd.crosstab(df['Género'],df['Medalla'], margins=True, margins_name='Total')

Medalla  bronze  gold  silver  Total
Género                              
Men         568   573     570   1711
Mixed        36    37      36    109
Women       344   349     352   1045
Total       948   959     958   2865

In [71]:
## Podemos mostrar las frecuencias 'normalizadas' (es decir una parte de la unidad). Se puede obtener respecto al total o con respecto a los subtotales
## de filas o columnas. Asignamos al parámetro normalize el valor True, la normalización se realiza con respecto al total de muestras
pd.crosstab(df['Género'],df['Medalla'], normalize=True)

Medalla    bronze      gold    silver
Género                               
Men      0.198255  0.200000  0.198953
Mixed    0.012565  0.012914  0.012565
Women    0.120070  0.121815  0.122862

Del total de las medallas entregadas: 

1.   19.8% fueron de bronce y entregas a hombres.
2.   12.2% fueron de plata y entregadas a mujeres.

En cada casilla podemos extraer información similar.


In [72]:
pd.crosstab(df['Género'],df['Medalla'], normalize=True, margins=True, margins_name= 'Total %')

Medalla    bronze      gold    silver   Total %
Género                                         
Men      0.198255  0.200000  0.198953  0.597208
Mixed    0.012565  0.012914  0.012565  0.038045
Women    0.120070  0.121815  0.122862  0.364747
Total %  0.330890  0.334729  0.334380  1.000000

In [73]:
pd.crosstab(df['Género'],df['Country'])

Country  Australia  Austria  Belarus  Belgium  Britain  Bulgaria  Canada  \
Género                                                                     
Men              5      143        8        4       13         1      82   
Mixed            0        5        0        1        2         0      11   
Women            7       70        7        0       11         5      77   

Country  China  Croatia  Czech Republic  ...  Soviet Union  Spain  Sweden  \
Género                                   ...                                
Men         12        5               9  ...           104      1     106   
Mixed        5        0               1  ...            20      0       0   
Women       36        6              14  ...            70      1      38   

Country  Switzerland  Ukraine  Unified Team  United States  Uzbekistan  \
Género                                                                   
Men               98        0             4            155           0   
Mixed              0        1             4             11           0   
Women             40        6            15            116           1   

Country  West Germany  Yugoslavia  
Género                             
Men                25           3  
Mixed               1           0  
Women              13           1  

[3 rows x 44 columns]

In [74]:
pais_genero = pd.DataFrame(pd.crosstab(df['Género'],df['Country']).unstack()).rename({0:'Conteo'},axis=1)
pais_genero.head(10)

Conteo
Country   Género        
Australia Men          5
          Mixed        0
          Women        7
Austria   Men        143
          Mixed        5
          Women       70
Belarus   Men          8
          Mixed        0
          Women        7
Belgium   Men          4

In [75]:
pais_genero.loc['Belgium']

Conteo
Género        
Men          4
Mixed        1
Women        0

### ***Archivos SPSS***

In [76]:
## Archivos de datos SPSS
df_spss = pd.read_spss('data/CARS2005.SAV')
df_spss

Brand                   Model  Minivan  Wagon  Pickup  \
0            TOYOTA                   Prius      0.0    0.0     0.0   
1             HONDA            Civic Hybrid      0.0    0.0     0.0   
2             HONDA            Civic Hybrid      0.0    0.0     0.0   
3             HONDA            Civic Hybrid      0.0    0.0     0.0   
4             HONDA            Civic Hybrid      0.0    0.0     0.0   
...             ...                     ...      ...    ...     ...   
1076     LAND ROVER             Range Rover      0.0    0.0     0.0   
1077  MERCEDES-BENZ                 G55 AMG      0.0    0.0     0.0   
1078      CHEVROLET  C2500 HD Silverado 2WD      0.0    0.0     1.0   
1079            GMC     C2500 HD Sierra 2WD      0.0    0.0     1.0   
1080        PORSCHE              Carrera GT      0.0    0.0     0.0   

      Automatic  EngineSize  Cylinders  CityMPG  HwyMPG  SUV AWD  Hybrid  
0           1.0         1.5        4.0     60.0    51.0   No  No     1.0  
1           1.0         1.3        4.0     48.0    47.0   No  No     1.0  
2           1.0         1.3        4.0     47.0    48.0   No  No     1.0  
3           0.0         1.3        4.0     46.0    51.0   No  No     1.0  
4           0.0         1.3        4.0     45.0    51.0   No  No     1.0  
...         ...         ...        ...      ...     ...  ...  ..     ...  
1076        1.0         4.4        8.0     12.0    16.0  Yes  No     0.0  
1077        1.0         5.4        8.0     12.0    14.0  Yes  No     0.0  
1078        1.0         6.0        8.0     11.0    13.0   No  No     0.0  
1079        1.0         6.0        8.0     11.0    13.0   No  No     0.0  
1080        0.0         5.7       10.0     10.0    16.0   No  No     0.0  

[1081 rows x 13 columns]

### ***Tablas de Wikipedia***

In [77]:
import requests
from bs4 import BeautifulSoup

url = 'https://es.wikipedia.org/wiki/Chicago_Bulls'
r = requests.get(url)
print(f'Response get --> {r}')
soup = BeautifulSoup(r.text, 'html.parser')
print(soup.prettify())

Response get --> <Response [200]>
<!DOCTYPE html>
<html class="client-nojs vector-feature-language-in-header-enabled vector-feature-language-in-main-page-header-disabled vector-feature-sticky-header-disabled vector-feature-page-tools-pinned-disabled vector-feature-toc-pinned-enabled vector-feature-main-menu-pinned-disabled vector-feature-limited-width-enabled vector-feature-limited-width-content-enabled vector-feature-zebra-design-disabled" dir="ltr" lang="es">
 <head>
  <meta charset="utf-8"/>
  <title>
   Chicago Bulls - Wikipedia, la enciclopedia libre
  </title>
  <script>
   document.documentElement.className="client-js vector-feature-language-in-header-enabled vector-feature-language-in-main-page-header-disabled vector-feature-sticky-header-disabled vector-feature-page-tools-pinned-disabled vector-feature-toc-pinned-enabled vector-feature-main-menu-pinned-disabled vector-feature-limited-width-enabled vector-feature-limited-width-content-enabled vector-feature-zebra-design-disable

### ***Tabla with BeautifulSoup.***

In [78]:
print(type(soup.find_all('table')))
tables = [x for x in soup.find_all('table')]

<class 'bs4.element.ResultSet'>


In [79]:
lista = []
for i in tables[0].find_all('tr'):
    try:
        i.find('td').get_text()
    except:
        lista.append(np.nan)
    else:
        lista.append(i.find('td').get_text())
#lista, len(lista)

In [80]:
lista1 = []
for i in tables[0].find_all('tr'):
    try:
        i.find('th').get_text()
    except:
        lista1.append(np.nan)
    else:
        lista1.append(i.find('th').get_text())
#lista1, len(lista1)

In [81]:
dicc = dict(zip(lista1, lista))
pd.DataFrame(dicc.items(), columns=['categoria', 'info'])

categoria                                               info
0            Chicago Bulls                                                NaN
1                      NaN                         [editar datos en Wikidata]
2          Datos generales                                                NaN
3                  Mascota                                   \nBenny the Bull
4                  Deporte                                       \nBaloncesto
5                Fundación                    \n16 de enero de 1966 (57 años)
6                  Colores           \nRojo, negro y blanco\n\n              
7           Propietario(s)                                 \n Jerry Reinsdorf
8               Presidente                               \n Michael Reinsdorf
9          Mánager general                                   \n Marc Eversley
10              Entrenador                                   \n Billy Donovan
11         Equipo afiliado                                 \nWindy City Bulls
12            Patrocinador                                         \nMotorola
13           Instalaciones                                                NaN
14        Estadio cubierto                                    \nUnited Center
15               Ubicación               \nChicago, Illinois,  Estados Unidos
16               Capacidad                                           \n20.917
17            Inauguración                   \n18 de agosto de 1994 (28 años)
18                Uniforme                                                NaN
19             Competición                                                NaN
20                    Liga                                              \nNBA
21             Conferencia                                 \nConferencia Este
22                División                                 \nDivisión Central
23                Palmarés                                                NaN
24                 Títulos           \n6 (1991, 1992, 1993, 1996, 1997, 1998)
25  Títulos de conferencia           \n6 (1991, 1992, 1993, 1996, 1997, 1998)
26     Títulos de división  \n9 (1975, 1991, 1992, 1993, 1996, 1997, 1998,...

In [82]:
result = pd.DataFrame(list(zip(lista1, lista)), columns=['col1', 'col2'])
result['col2'] = result['col2'].apply(lambda x: str(x).replace('\n',""))
result = result.dropna(subset=['col1'])
result

col1                                               col2
0            Chicago Bulls                                                nan
2          Datos generales                                                nan
3                  Mascota                                     Benny the Bull
4                  Deporte                                         Baloncesto
5                Fundación                      16 de enero de 1966 (57 años)
6                  Colores                 Rojo, negro y blanco              
7           Propietario(s)                                    Jerry Reinsdorf
8               Presidente                                  Michael Reinsdorf
9          Mánager general                                      Marc Eversley
10              Entrenador                                      Billy Donovan
11         Equipo afiliado                                   Windy City Bulls
12            Patrocinador                                           Motorola
13           Instalaciones                                                nan
14        Estadio cubierto                                      United Center
15               Ubicación                 Chicago, Illinois,  Estados Unidos
16               Capacidad                                             20.917
17            Inauguración                     18 de agosto de 1994 (28 años)
18                Uniforme                                                nan
21             Competición                                                nan
22                    Liga                                                NBA
23             Conferencia                                   Conferencia Este
24                División                                   División Central
25                Palmarés                                                nan
26                 Títulos             6 (1991, 1992, 1993, 1996, 1997, 1998)
27  Títulos de conferencia             6 (1991, 1992, 1993, 1996, 1997, 1998)
28     Títulos de división  9 (1975, 1991, 1992, 1993, 1996, 1997, 1998, 2...

In [84]:
## Tablas de wikipedia
df_wiki = pd.read_html('https://es.wikipedia.org/wiki/Chicago_Bulls', )
for i, df in enumerate(df_wiki):
    print(i+1)
    display(df.head(5))

1


Chicago Bulls                Chicago Bulls.1  \
0              NaN                            NaN   
1  Datos generales                Datos generales   
2          Mascota                 Benny the Bull   
3          Deporte                     Baloncesto   
4        Fundación  16 de enero de 1966 (57 años)   

                 Chicago Bulls.2  
0                            NaN  
1                Datos generales  
2                 Benny the Bull  
3                     Baloncesto  
4  16 de enero de 1966 (57 años)

2


Temporada              G              P              %  \
0  Chicago Bulls  Chicago Bulls  Chicago Bulls  Chicago Bulls   
1        1966-67             33             49           .407   
2        1967-68             29             53           .354   
3        1968-69             33             49           .402   
4        1969-70             39             43           .476   

                     Play-offs               Resultado  
0                Chicago Bulls           Chicago Bulls  
1  Pierde Semifinales División  St. Louis 3, Chicago 0  
2  Pierde Semifinales División  LA Lakers 4, Chicago 1  
3                          NaN                     NaN  
4  Pierde Semifinales División    Atlanta 4, Chicago 1

3


Chicago Bulls 2022-23  \
                                           Jugadores   
0  N.º Nac. Pos. Nombre Altura Peso Origen 2 B Ba...   
1                                                N.º   
2                                                  2   
3                                                 13   
4                                                  6   

                                                     Unnamed: 2_level_0  \
                                        Entrenadores Unnamed: 2_level_1   
0  Entrenador Billy Donovan Ayudante(s) John Brya...                NaN   
1                                               Nac.               Pos.   
2                                                NaN                  B   
3                                                NaN                  P   
4                                                NaN                  B   

  Unnamed: 3_level_0   Unnamed: 4_level_0 Unnamed: 5_level_0  \
  Unnamed: 3_level_1   Unnamed: 4_level_1 Unnamed: 5_level_1   
0                NaN                  NaN                NaN   
1             Nombre               Altura               Peso   
2        Ball, Lonzo   1.98 m (6 ft 6 in)     86 kg (190 lb)   
3      Bradley, Tony  2.08 m (6 ft 10 in)    112 kg (248 lb)   
4       Caruso, Alex   1.93 m (6 ft 4 in)     84 kg (186 lb)   

  Unnamed: 6_level_0  
  Unnamed: 6_level_1  
0                NaN  
1             Origen  
2               UCLA  
3     North Carolina  
4          Texas A&M

4


N.º  Nac. Pos.          Nombre               Altura             Peso  \
0    2   NaN    B     Ball, Lonzo   1.98 m (6 ft 6 in)   86 kg (190 lb)   
1   13   NaN    P   Bradley, Tony  2.08 m (6 ft 10 in)  112 kg (248 lb)   
2    6   NaN    B    Caruso, Alex   1.93 m (6 ft 4 in)   84 kg (186 lb)   
3   11   NaN    E  DeRozan, DeMar   1.98 m (6 ft 6 in)  100 kg (220 lb)   
4   12   NaN  B-E    Dosunmu, Ayo   1.93 m (6 ft 4 in)   91 kg (200 lb)   

                Origen  
0                 UCLA  
1       North Carolina  
2            Texas A&M  
3  Southern California  
4             Illinois

5


0   1                     2   3   \
0  1966 – 1976 Jerry Sloan NaN  1968 – 1976 Bob Love NaN   

                                       4   5   \
0  1984 – 1993 1995 – 1998 Michael Jordan NaN   

                                       6   7                         8   9   \
0  1987 – 1998 2003 – 2004 Scottie Pippen NaN  1989 – 1998 Phil Jackson NaN   

                                    10  11                        12  13  14  
0  1966 – 1968 1977 – 2009 Johnny Kerr NaN  1985 – 2003 Jerry Krause NaN NaN

6


#         Jugador                      Temporadas Promedio Partidos   Total
0  1  Michael Jordan  1984 - 1993 / 1994 - 1998 (13)      314      932  29.277
1  2  Scottie Pippen                1987 - 1998 (11)      177      856  15.123
2  3        Bob Love                 1968 - 1977 (9)      213      592  12.623
3  4       Luol Deng                 2004 - 2013 (9)      161      637  10.286
4  5     Jerry Sloan                1966 - 1976 (10)      147      696  10.233

7


#         Jugador                      Temporadas Promedio Partidos  Total
0  1  Michael Jordan  1984 - 1993 / 1994 - 1998 (13)       63      932  5.836
1  2  Tom Boerwinkle                1968 - 1978 (10)       90      636  5.745
2  3  Scottie Pippen                1987 - 1998 (11)       67      856  5.726
3  4     Joakim Noah                 2007 - 2016 (9)       94      572   5387
4  5     Jerry Sloan                1966 - 1976 (10)       77      696  5.385

8


#         Jugador                      Temporadas Promedio Partidos  Total
0  1  Michael Jordan  1984 - 1993 / 1994 - 1998 (13)       54      932  5.012
1  2  Scottie Pippen                1987 - 1998 (11)       53      856  4.494
2  3    Kirk Hinrich  2003 - 2010 / 2012 - 2016 (11)       50      748  3.811
3  4    Derrick Rose                 2008 - 2016 (7)       62      406  2.516
4  5   Norm Van Lier                 1971 - 1978 (7)       69      535  3.676

9


GM    Periodo
0    Dick Klein  1967–1969
1  Pat Williams  1969–1973
2    Dick Motta  1973–1976
3  Jerry Krause       1976
4    Jon Kovler  1976–1978

10


0  \
0  MVP de la Temporada Michael Jordan - 1988, 199...   

                                                   1  \
0  Mejor quinteto de la NBA Michael Jordan - 1987...   

                                                   2  
0  Mejor quinteto de rookies Erwin Mueller - 1967...

11


0                                                  1
0  Control de autoridades  Proyectos Wikimedia  Datos: Q128109  Multimedi...